slides.typ

```diff
@@ -1,23 +1,222 @@
+        }
+
+        lower-okay and upper-okay
+    } else {
+        panic("you may only provide a single integer, an array of integers, or a string")
+    }
+}
+
+#let _last-required-subslide(visible-subslides) = {
+    if type(visible-subslides) == "integer" {
+        visible-subslides
+    } else if type(visible-subslides) == "array" {
+        calc.max(..visible-subslides.map(s => _last-required-subslide(s)))
+    } else if type(visible-subslides) == "string" {
+        let parts = _parse-subslide-indices(visible-subslides)
+        _last-required-subslide(parts)
+    } else if type(visible-subslides) == "dictionary" {
+        let last = 0
+        if "beginning" in visible-subslides {
+            last = calc.max(last, visible-subslides.beginning)
+        }
+        if "until" in visible-subslides {
+            last = calc.max(last, visible-subslides.until)
+        }
+        last
+    } else {
+        panic("you may only provide a single integer, an array of integers, or a string")
+    }
+}
+
+#let _conditional-display(visible-subslides, reserve-space, mode, body) = {
+    repetitions.update(rep => calc.max(rep, _last-required-subslide(visible-subslides)))
+    locate( loc => {
+        if _check-visible(subslide.at(loc).first(), visible-subslides) {
+            body
+        } else if reserve-space {
+            _slides-cover(mode, body)
+        }
+    })
+}
+
+#let uncover(visible-subslides, mode: "invisible", body) = {
+    _conditional-display(visible-subslides, true, mode, body)
+}
+
+#let only(visible-subslides, body) = {
+    _conditional-display(visible-subslides, false, "doesn't even matter", body)
+}
+
+#let one-by-one(start: 1, mode: "invisible", ..children) = {
+    repetitions.update(rep => calc.max(rep, start + children.pos().len() - 1))
+    for (idx, child) in children.pos().enumerate() {
+        uncover((beginning: start + idx), mode: mode, child)
+    }
 }
 
+#let alternatives(start: 1, position: bottom + left, ..children) = {
+    repetitions.update(rep => calc.max(rep, start + children.pos().len() - 1))
+    style(styles => {
+        let sizes = children.pos().map(c => measure(c, styles))
+        let max-width = calc.max(..sizes.map(sz => sz.width))
+        let max-height = calc.max(..sizes.map(sz => sz.height))
+        for (idx, child) in children.pos().enumerate() {
+            only(start + idx, box(
+                width: max-width,
+                height: max-height,
+                align(position, child)
+            ))
+        }
+    })
+}
+
+#let line-by-line(start: 1, mode: "invisible", body) = {
+    let items = if repr(body.func()) == "sequence" {
+        body.children
+    } else {
+        ( body, )
+    }
+
+    let idx = start
+    for item in items {
+        if repr(item.func()) != "space" {
+            uncover((beginning: idx), mode: mode, item)
+            idx += 1
+        } else {
+            item
+        }
+    }
+}
+
+#let pause(mode: "invisible", body) = {
```

Comment:
I have trouble using this here:
```
#import "slides-pr.typ": *

#show: slides.with(
    authors: "",
)

#new-section("Introduction")

#slide(title:"slide 1")[
  == header 1
    #only(1)[A small text 1.]
    #show: pause

  == header 2
    #only(2)[A small text 2.]
    #show: pause

  == header 3
    #only(3)[A small text 3.]
    #show: pause
]
```
Header 3 appears with Header 2 and "small text 3" never appears at all. Am I misusing it?

Reply:
Interesting... I will look into that.

Reply:
Apparently the problem here is the last `#show: pause`. If you remove that, everything works nicely. As to _why_ that is the problem... I don't know yet.

Reply:
It seems the 3rd `#show: pause` is one root cause. This example don't work with 4 subslides and not final `pause` call:
```
...
#slide(title:"slide 1")[
  == header 1
    #only(1)[A small text 1.]
  #show: pause

  == header 2
    #only(2)[A small text 2.]
  #show: pause

  == header 3
    #only(3)[A small text 3.]
  #show: pause

  == header 4
    #only(4)[A small text 4.]
]
```

Reply:
Update: The root cause seems to be related to the `hide(body)` call in `_slides-cover(mode, body)`. If I switch the default `pause` mode to `"transparent"` it appears to work as intended.

Reply:
Hmm yes I think you're on to something...